libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,296 @@
+                    if (locationBuffer.size > 0) {
+                        locationBuffer.removeAt(0)
+                    }
+                }
+                LocationBufferCommands.BUFFER_CLEAR -> {
+                    locationBuffer.clear()
+                }
+                LocationBufferCommands.BUFFER_READ -> {
+                    locationBuffer.forEach { location ->
+                        locationAction.predicate(location)
+                    }
+                }
+            }
+        }
+
+        /**
+         * This monitors the life time of the job that handles the channels. If that job is terminated, we cleanup.
+         */
+        jobControl.scope.launch {
+            select<Unit> {
+                cleanupJob.onJoin {
+                    locationEngine.removeLocationUpdates(locationCallback)
+                }
+            }
+        }
+    }
+
+    private suspend fun getDatePartitionedLocations(predicate: (Location) -> Boolean): Array<Location> {
+        val retVal = mutableListOf<Location>()
+        channelLocationBuffer.send(LocationBufferControl(LocationBufferCommands.BUFFER_READ, null) { location ->
+            if (predicate(location)) {
+                retVal.add(location)
+            }
+        })
+        return retVal.toTypedArray()
+    }
+
+    private fun parseRouteProgressAsync(): Deferred<TelemetryStep> {
+        return jobControl.scope.async {
+            val telemetryStep = TelemetryStep() // Default initialization
+            val routeProgress = channelOnRouteProgress.receive()
+            routeProgress.currentLegProgress()?.upcomingStep()?.let { legStep ->
+                telemetryStep.distance = legStep.distance().toInt()
+                telemetryStep.distanceRemaining = routeProgress.distanceRemaining().toInt()
+                telemetryStep.duration = legStep.duration().toInt()
+                telemetryStep.durationRemaining = routeProgress.durationRemaining().toInt()
+                telemetryStep.previousModifier = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.maneuver()?.modifier()
+                        ?: ""
+                telemetryStep.previousName = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.name()
+                        ?: ""
+                telemetryStep.previousType = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.maneuver()?.type()
+                        ?: ""
+                telemetryStep.upcomingType = legStep.maneuver().type() ?: ""
+                telemetryStep.upcomingModifier = legStep.maneuver().modifier() ?: ""
+                telemetryStep.upcomingName = legStep.name() ?: ""
+            }
+            telemetryStep
+        }
+    }
+
+    /**
+     * TODO:OZ add one of these for each telemetry event
+     * Generates a well-formed Feedback event.
+     */
+    private suspend fun populateUserFeedbackEvent(event: TelemetryEventFeedback): TelemetryEventInterface {
```

Comment:
Currently (legacy) `FeedbackEvent`s are only be sent when in `Active Guidance`.

Reply:
👍 changed to allow posting user feedback events only during an active session. @Guardiola31337 although we discussed adding an additional check to see if there is a primary route available, and the MapboxTelemetry code says as much in the comments, I did not see this check in the code. So I am refraining from adding this check in telemetry until the same check is made in MapboxNavigation